### PR TITLE
Semantic snippets: Make postfix `for` and `forr` snippets available for types with `Length` or `Count` property

### PR DESCRIPTION
**Pull request by DoctorKrolic**
It is a very common pattern to iterate through a collection by indices in case these indices are needed in the loop. It was previously possible to make a for/reversed for loop over a collection like this, but it required to type full `Length`/`Count` property and activating snippet after that. In this PR I simplified this, so you can just use `for`/`forr` on a type, which has either suitable `Length` or `Count` property:
devenv_WS81Uosau0

### CODE REVIEW

**Comment by DoctorKrolic** [comment]
@akhera99 PTAL

**Review by akhera99 — APPROVED** [review]
Looks really good! Thanks for the contribution!

File: src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs [hunk 1]
```diff
@@ -48,19 +47,42 @@ protected override SyntaxNode GenerateStatement(SyntaxGenerator generator, Synta
 
             TypeSyntax? iteratorTypeSyntax = null;
 
+            var inlineExpression = inlineExpressionInfo?.Node.WithoutLeadingTrivia();
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs:50]: could you extract this all to a helper local function?

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs:50]: specifically, i really hate having to look ata ll the flow paths and see how they interact with each other :)

File: src/Features/Core/Portable/Snippets/SnippetProviders/AbstractForLoopSnippetProvider.cs [hunk 2]
```diff
@@ -11,9 +12,36 @@ namespace Microsoft.CodeAnalysis.Snippets.SnippetProviders
...
+        protected static IPropertySymbol? FindLengthProperty(ITypeSymbol type, Compilation compilation)
+            => FindAccessibleIntegerProperty(type, compilation, "Length");
+
+        protected static IPropertySymbol? FindCountProperty(ITypeSymbol type, Compilation compilation)
+            => FindAccessibleIntegerProperty(type, compilation, "Count");
+
+        private static IPropertySymbol? FindAccessibleIntegerProperty(ITypeSymbol type, Compilation compilation, string propertyName)
+        {
+            return type
+                .GetAccessibleMembersInThisAndBaseTypes<IPropertySymbol>(compilation.Assembly)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Snippets/SnippetProviders/AbstractForLoopSnippetProvider.cs:43]: is there a GetAccessibleMembers that takes a name?  would be a lot better than getting all members (Expensive) and having to iterate over all of them (expensive).

File: src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs [hunk 3]
```diff
@@ -45,21 +44,7 @@ protected override SyntaxNode GenerateStatement(SyntaxGenerator generator, Synta
...
-            }
-            else
-            {
-                var inlineExpressionType = inlineExpressionInfo.TypeInfo.Type;
-                Debug.Assert(inlineExpressionType is not null && (inlineExpressionType.IsIntegralType() || inlineExpressionType.IsNativeIntegerType));
-                iteratorTypeSyntax = inlineExpressionType.GenerateTypeSyntax();
-            }
-
-            var inlineExpression = inlineExpressionInfo?.Node.WithoutLeadingTrivia();
+            GetLoopHeaderParts(generator, inlineExpressionInfo, compilation, out var iteratorTypeSyntax, out var inlineExpression);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs:47]:
```suggestion
            var (iteratorTypeSyntax, inlineExpression) = GetLoopHeaderParts(generator, inlineExpressionInfo, compilation);
```

File: src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs [hunk 4]
```diff
@@ -76,6 +61,48 @@ protected override SyntaxNode GenerateStatement(SyntaxGenerator generator, Synta
                 SyntaxFactory.SingletonSeparatedList<ExpressionSyntax>(
                     SyntaxFactory.PostfixUnaryExpression(IncrementorKind, indexVariable)),
                 SyntaxFactory.Block());
+
+            static (TypeSyntax iteratorTypeSyntax, SyntaxNode? inlineExpression) GetLoopHeaderParts(SyntaxGenerator generator, InlineExpressionInfo? inlineExpressionInfo, Compilation compilation)
+            {
+                TypeSyntax iteratorTypeSyntax;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs:67]:
```suggestion
```

File: src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs [hunk 5]
```diff
@@ -76,6 +61,48 @@ protected override SyntaxNode GenerateStatement(SyntaxGenerator generator, Synta
...
+                var inlineExpression = inlineExpressionInfo?.Node.WithoutLeadingTrivia();
+
+                if (inlineExpressionInfo is null)
+                {
+                    iteratorTypeSyntax = compilation.GetSpecialType(SpecialType.System_Int32).GenerateTypeSyntax();
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs:72]:
```suggestion
                    return (compilation.GetSpecialType(SpecialType.System_Int32).GenerateTypeSyntax(), inlineExpression);
```

File: src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs [hunk 6]
```diff
@@ -76,6 +61,48 @@ protected override SyntaxNode GenerateStatement(SyntaxGenerator generator, Synta
...
+                }
+                else
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs:74]: consider removing else block.

File: src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs [hunk 7]
```diff
@@ -76,6 +61,48 @@ protected override SyntaxNode GenerateStatement(SyntaxGenerator generator, Synta
...
+                {
+                    var inlineExpressionType = inlineExpressionInfo.TypeInfo.Type;
+                    Debug.Assert(inlineExpressionType is not null);
+
+                    if (IsSuitableIntegerType(inlineExpressionType))
+                    {
+                        iteratorTypeSyntax = inlineExpressionType.GenerateTypeSyntax();
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs:81]:
```suggestion
                        return (inlineExpressionType.GenerateTypeSyntax(), inlineExpression);
```

File: src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs [hunk 8]
```diff
@@ -76,6 +61,48 @@ protected override SyntaxNode GenerateStatement(SyntaxGenerator generator, Synta
...
+                    }
+                    else
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs:83]: consider removing this else.

File: src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs [hunk 9]
```diff
@@ -76,6 +61,48 @@ protected override SyntaxNode GenerateStatement(SyntaxGenerator generator, Synta
...
+                        {
+                            iteratorTypeSyntax = lengthProperty.Type.GenerateTypeSyntax();
+                            inlineExpression = generator.MemberAccessExpression(inlineExpression, "Length");
+                        }
+                        else
+                        {
+                            Debug.Assert(countProperty is not null);
+                            iteratorTypeSyntax = countProperty.Type.GenerateTypeSyntax();
+                            inlineExpression = generator.MemberAccessExpression(inlineExpression, "Count");
+                        }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs:100]: feels collapsible into a conditional-expr.

File: src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs [hunk 10]
```diff
@@ -76,6 +61,48 @@ protected override SyntaxNode GenerateStatement(SyntaxGenerator generator, Synta
...
+                    }
+                }
+
+                return (iteratorTypeSyntax, inlineExpression);
+            }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Snippets/AbstractCSharpForLoopSnippetProvider.cs:105]:
```suggestion
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs [hunk 11]
```diff
@@ -450,6 +450,16 @@ public static bool CanSupportCollectionInitializer(this ITypeSymbol typeSymbol,
...
 
+        public static ImmutableArray<T> GetAccessibleMembersInThisAndBaseTypes<T>(this ITypeSymbol? containingType, string memberName, ISymbol within) where T : class, ISymbol
+        {
+            if (containingType == null)
+            {
+                return ImmutableArray<T>.Empty;
+            }
+
+            return containingType.GetBaseTypesAndThis().SelectAccessibleMembers<T>(memberName, within).ToImmutableArray();
+        }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ITypeSymbolExtensions.cs:461]:
```suggestion
        => containingType == null
            ? ImmutableArray<T>.Empty
            : containingType.GetBaseTypesAndThis().SelectAccessibleMembers<T>(memberName, within).ToImmutableArray();
```

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Note: i'm happy to make these changes if you'd prefer i take thsi off your shoulders.

**Comment by DoctorKrolic** [comment]
> i'm happy to make these changes if you'd prefer i take thsi off your shoulders

Go ahead! It's late evening for me and I've just finished another complicated PR, so thinking is somewhat hard at that point)

**Comment by DoctorKrolic** [comment]
@CyrusNajmabadi I resolved merge conflict with your latest collection expressions pass. I think, this is ready for merge when CI finishes

**Pull request merged**